pulsar-broker-auth-sasl/src/main/java/org/apache/pulsar/broker/authentication/AuthenticationProviderSasl.java

```diff
@@ -0,0 +1,108 @@
+import java.util.regex.Pattern;
+import java.util.regex.PatternSyntaxException;
+import javax.naming.AuthenticationException;
+import javax.security.auth.login.LoginException;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.pulsar.broker.ServiceConfiguration;
+import org.apache.pulsar.common.sasl.JAASCredentialsContainer;
+import org.apache.pulsar.common.sasl.SaslConstants;
+
+@Slf4j
+public class AuthenticationProviderSasl implements AuthenticationProvider {
+
+    private Pattern allowedIdsPattern;
+    private Map<String, String> configuration;
+
+    private JAASCredentialsContainer jaasCredentialsContainer;
+    private String loginContextName;
+
+    @Override
+    public void initialize(ServiceConfiguration config) throws IOException {
+        this.configuration = Maps.newHashMap();
+        final String allowedIdsPatternRegExp = config.getSaslJaasClientAllowedIds();
+        configuration.put(JAAS_CLIENT_ALLOWED_IDS, allowedIdsPatternRegExp);
+        configuration.put(JAAS_BROKER_SECTION_NAME, config.getSaslJaasBrokerSectionName());
+        configuration.put(KINIT_COMMAND, config.getKinitCommand());
+
+        try {
+            this.allowedIdsPattern = Pattern.compile(allowedIdsPatternRegExp);
+        } catch (PatternSyntaxException error) {
+            log.error("Invalid regular expression for id " + allowedIdsPatternRegExp, error);
+            throw new IOException(error);
+        }
+
+        loginContextName = config.getSaslJaasBrokerSectionName();
+        if (jaasCredentialsContainer == null) {
+            log.info("JAAS loginContext is: {}." , loginContextName);
+            try {
+                jaasCredentialsContainer = new JAASCredentialsContainer(
+                    loginContextName,
+                    new PulsarSaslServer.SaslServerCallbackHandler(allowedIdsPattern),
+                    configuration);
+            } catch (LoginException e) {
+                log.error("JAAS login in broker failed: {}" , e);
+                throw new IOException(e);
+            }
+        }
+    }
+
+    @Override
+    public AuthenticationDataSource getAuthDataSource() throws IOException {
+        try {
+            PulsarSaslServer saslServer = new PulsarSaslServer(jaasCredentialsContainer.getSubject(), allowedIdsPattern);
+            return new SaslAuthenticationDataSource(saslServer);
+        } catch (Throwable t) {
+            log.error("Failed create sasl server: {}" , t);
+            throw new IOException(t);
+        }
+    }
+
+    // will get authenticated client id like: "client/host.name.client@EXAMPLE.COM"
+    @Override
+    public String authenticate(AuthenticationDataSource authData) throws AuthenticationException {
+        if (authData instanceof SaslAuthenticationDataSource) {
+            return authData.getAuthorizationID();
+        } else {
+            throw new AuthenticationException("Not support get authDate from outside sasl.");
+        }
+        // TODO: for http
```

Comment:
actually do we need this TODO comment here?

Reply:
Thanks, will remove it